crypto/rand/rand_unix.c

```diff
@@ -119,26 +134,89 @@ size_t rand_pool_acquire_entropy(RAND_POOL *pool)
 #   error "Seeding uses urandom but DEVRANDOM is not configured"
 #  endif
 
+#  if defined(__GLIBC__) && defined(__GLIBC_PREREQ)
+#   if __GLIBC_PREREQ(2, 25)
+#    define OPENSSL_HAVE_GETRANDOM
+#   endif
+#  endif
+
+#  if (defined(__FreeBSD__) && __FreeBSD_version >= 1200061)
+#   define OPENSSL_HAVE_GETRANDOM
+#  endif
+
+#  if defined(OPENSSL_HAVE_GETRANDOM)
+#   include <sys/random.h>
+#  endif
+
 #  if defined(OPENSSL_RAND_SEED_OS)
 #   if !defined(DEVRANDOM)
 #    error "OS seeding requires DEVRANDOM to be configured"
 #   endif
+#   define OPENSSL_RAND_SEED_GETRANDOM
 #   define OPENSSL_RAND_SEED_DEVRANDOM
-#   if defined(__GLIBC__) && defined(__GLIBC_PREREQ)
-#    if __GLIBC_PREREQ(2, 25)
-#     define OPENSSL_RAND_SEED_GETRANDOM
-#    endif
-#   endif
-#  endif
-
-#  ifdef OPENSSL_RAND_SEED_GETRANDOM
-#   include <sys/random.h>
 #  endif
 
 #  if defined(OPENSSL_RAND_SEED_LIBRANDOM)
 #   error "librandom not (yet) supported"
 #  endif
 
+#  if defined(__FreeBSD__) && defined(KERN_ARND)
+/*
+ * sysctl_random(): Use sysctl() to read a random number from the kernel
+ * Returns the size on success, 0 on failure.
+ */
+static size_t sysctl_random(void *buf, size_t buflen)
+{
+    int mib[2];
+    size_t done = 0;
+    size_t len;
+
+    /* The sysctl returns longs. */
+    if (buflen % sizeof(long) != 0)
+        return 0;
+
+    mib[0] = CTL_KERN;
+    mib[1] = KERN_ARND;
+
+    do {
+        len = buflen;
+        if (sysctl(mib, 2, buf, &len, NULL, 0) == -1)
```

Comment:
It's malloc()'d. I'm not sure why you think there is an alignment issue, or how I need to resolve it.

Reply:
Well, a few lines up it's claimed that the data is returned in multiples of 'long', as if the long type was used internally.  The compiler will not complain at us if we pass a ```void*``` where a ```long*``` is expected, leaving it to the programmer to ensure that the pointer value in question has sufficient alignment for the type in question (since formally, we're supposed to only actually be using a pointer to the type that we end up dereferencing it as).  As you note, this buffer is going to start its life as a malloc output, which is not intrinsically type long.  That said, malloc output is required to be sufficiently aligned for use as any native type, so there should not be a problem in practice at runtime.  (There would maybe be issues if this function was exposed to user-supplied argument values, but I did not attempt to trace if that is possible.)